gax/src/main/java/com/google/api/gax/rpc/FirstElementResponseObserver.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2017, Google LLC All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.rpc;
+
+import com.google.api.core.AbstractApiFuture;
+import com.google.api.core.ApiFuture;
+
+/**
+ * A {@link ResponseObserver} that wraps a future. The future will resolved upon receiving the first
+ * element or exhausting the streaming.
+ *
+ * <p>Package-private for internal use.
+ *
+ * @param <ResponseT> The type of the element in the stream.
+ */
+class FirstElementResponseObserver<ResponseT> implements ResponseObserver<ResponseT> {
+  private final MyFuture future = new MyFuture();
+  private StreamController controller;
+
+  @Override
+  public void onStart(StreamController controller) {
+    // NOTE: the call is started before the future is exposed to the caller
+    this.controller = controller;
+
+    controller.disableAutoInboundFlowControl();
+    controller.request(1);
+  }
+
+  @Override
+  public void onResponse(ResponseT response) {
+    future.set(response);
+    controller.cancel();
+  }
+
+  @Override
+  public void onError(Throwable t) {
+    future.setException(t);
```

Comment:
What will happen if:

1. We get a response, and the future is set
2. We next get an error, so this attempts to set an exception on the future

Maybe my question boils down to this: Are we sure that if we cancel the controller, we won't get any more calls?

Reply:
Yes, a future's internal state can only be set once, from guava's javadocs:
Sets the result of this Future unless this Future has already been cancelled or set (including set asynchronously). 

But to be sure, I added a test for that behavior